Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp

```diff
@@ -96,56 +95,42 @@ void HeliumAnalyserEfficiency::init() {
   setPropertyGroup(PropertyNames::OUTPUT_FIT_PARAMS, PropertyNames::GROUP_OUTPUTS);
 }
 
+void validateInputWorkspace(MatrixWorkspace_sptr const &workspace, std::map<std::string, std::string> &errorList) {
+  Kernel::Unit_const_sptr unit = workspace->getAxis(0)->unit();
+  if (unit->unitID() != "Wavelength") {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be in units of Wavelength.";
+    return;
+  }
+}
+
 /**
  * Tests that the inputs are all valid
  * @return A map containing the incorrect workspace
  * properties and an error message
  */
 std::map<std::string, std::string> HeliumAnalyserEfficiency::validateInputs() {
   std::map<std::string, std::string> errorList;
-  const std::string inputWorkspaceName = getProperty(PropertyNames::INPUT_WORKSPACE);
-  if (!AnalysisDataService::Instance().doesExist(inputWorkspaceName)) {
+  const WorkspaceGroup_sptr wsGroup = getProperty(PropertyNames::INPUT_WORKSPACE);
+
+  if (wsGroup == nullptr) {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace is not a group workspace";
+  } else if (wsGroup->size() != 4) {
     errorList[PropertyNames::INPUT_WORKSPACE] =
-        "The input workspace " + inputWorkspaceName + " does not exist in the ADS.";
+        "The input group workspace must have four periods corresponding to the four spin configurations.";
   } else {
-    const auto ws = AnalysisDataService::Instance().retrieve(getProperty(PropertyNames::INPUT_WORKSPACE));
-    if (!ws->isGroup()) {
-      errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace is not a group workspace";
-    } else {
-      const auto wsGroup = std::dynamic_pointer_cast<WorkspaceGroup>(ws);
-      if (wsGroup->size() != 4) {
-        errorList[PropertyNames::INPUT_WORKSPACE] =
-            "The input group workspace must have four periods corresponding to the four spin configurations.";
-      }
+    for (size_t i = 0; i < wsGroup->size(); ++i) {
+      MatrixWorkspace_sptr const stateWs = std::dynamic_pointer_cast<MatrixWorkspace>(wsGroup->getItem(i));
```

Comment:
This is very much an edge case that is unlikely to be encountered in practice, but I can cause a crash by running this script:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

wsTPara = CreateEmptyTableWorkspace()
wsTPara1 = CreateEmptyTableWorkspace()
wsTAnti = CreateEmptyTableWorkspace()
wsTAnti1 = CreateEmptyTableWorkspace()

grp = GroupWorkspaces([wsTPara,wsTAnti,wsTPara1,wsTAnti1])
eff = HeliumAnalyserEfficiency(grp, SpinStates='11,01,00,10')
```

This creates an input group of table workspaces, rather than the expected input group of matrix workspaces. On line 122 of the algorithm where we are getting the individual workspace from the group and casting it to a `MatrixWorkspace`, the variable `stateWs` will be a `nullptr` if the type of the workspace is not a `MatrixWorkspace`. When we use this later inside `validateInputWorkspace` we then get a crash when we try to get the axis from it on line 99). To avoid the crash we should add a line to the start of the `validateInputWorkspace` method that checks for a `nullptr` and returns straight away (with appropriate error message) if one is found. So the `validateInputWorkspace` method would be refactored to be:

```
void validateInputWorkspace(MatrixWorkspace_sptr const &workspace, std::map<std::string, std::string> &errorList) {
  if (workspace == nullptr) {
    errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be of type MatrixWorkspace.";
    return;
  }

  Kernel::Unit_const_sptr unit = workspace->getAxis(0)->unit();
  if (unit->unitID() != "Wavelength") {
    errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be in units of Wavelength.";
  }
}
```